ast/src/parsed/pattern.rs

```diff
@@ -0,0 +1,432 @@
+        }
+    }
+
+    /// Specialize a pattern based on a "constructor" pattern passed as a parameter.
+    /// Based on https://doc.rust-lang.org/nightly/nightly-rustc/rustc_pattern_analysis/usefulness/index.html#specialization.
+    /// If the constructor is shared with the pattern, specialize the pattern removing
+    /// "one layer" (the constructor) and returning internal patterns.
+    /// Return None if the pattern can't be specialized using the constructor.
+    pub fn specialize(&self, constructor: &Self) -> Option<PatternTuple> {
+        match (constructor, self) {
+            (_, Pattern::CatchAll) => Some(PatternTuple { patterns: vec![] }),
+            (Pattern::CatchAll, _) => None,
+            (Pattern::Number(y), Pattern::Number(x)) => {
+                (y == x).then_some(PatternTuple { patterns: vec![] })
+            }
+            (Pattern::String(r), Pattern::String(s)) => {
+                (r == s).then_some(PatternTuple { patterns: vec![] })
+            }
+
+            (Pattern::Tuple(cons_patterns), Pattern::Tuple(patterns))
+                if cons_patterns.len() == patterns.len() =>
+            {
+                Some(PatternTuple {
+                    patterns: patterns.to_vec(),
+                })
+            }
+            (Pattern::Array(cons_patterns), Pattern::Array(patterns)) => {
+                let max_len = cons_patterns.len();
+                let mut constructors = Vec::new();
+                if patterns.contains(&Pattern::Ellipsis) {
+                    let mut expanded_items = patterns.clone();
+                    let ellipsis_index = patterns
+                        .iter()
+                        .position(|p| p == &Pattern::Ellipsis)
+                        .unwrap();
+                    let ellipsis_len = max_len - patterns.len() + 1;
+
+                    expanded_items.remove(ellipsis_index);
+
+                    expanded_items.splice(
+                        ellipsis_index..ellipsis_index,
+                        std::iter::repeat(Pattern::CatchAll).take(ellipsis_len),
+                    );
+
+                    constructors.extend(expanded_items);
+                } else {
+                    constructors = patterns.to_vec();
+                }
+
+                Some(PatternTuple {
+                    patterns: constructors,
+                })
+            }
+            (_, Pattern::Variable(var)) => Some(PatternTuple {
+                patterns: [].to_vec(),
+            }),
+
+            (Pattern::Enum(path1, patterns1), Pattern::Enum(path2, patterns2)) => {
+                if path1 != path2 {
+                    return None;
+                }
+                match (patterns1, patterns2) {
+                    (_, Some(_)) => {
+                        let pats = patterns2.as_ref().unwrap();
+                        Some(PatternTuple {
+                            patterns: pats.clone(),
+                        })
```

Comment:
```suggestion
                        Some(PatternTuple {
                            patterns: patterns2.as_ref().unwrap().clone(),
                        })
```
(or maybe even `.cloned()` cold work here?)

Reply:
Oh wait, why don't you use `Some(patterns)` instead of `Some(_)` one line above?